File: templates/mcp-server/src/schemas.ts
/**
 * GraphRAG Helper Plugin - Schemas
 *
 * Zod schemas for GraphRAG operations with automatic JSON Schema generation.
 */

import { z } from 'zod';

// ============================================================================
// Store Memory Tool Schemas
// ============================================================================

export const StoreMemoryInputSchema = z.object({
  content: z.string().min(1).max(10000).describe('The memory content to store'),
  tags: z.array(z.string()).max(10).optional().describe('Tags for categorization'),
  metadata: z.record(z.any()).optional().describe('Additional metadata'),
  importance: z.number().min(0).max(1).default(0.5).describe('Importance score (0-1)'),
});
export type StoreMemoryInput = z.infer<typeof StoreMemoryInputSchema>;

export const StoreMemoryOutputSchema = z.object({
  id: z.string().describe('Unique memory ID'),
  stored: z.boolean().describe('Whether storage was successful'),
  timestamp: z.string().datetime().describe('Storage timestamp'),
  embedding: z.object({
    model: z.string().describe('Embedding model used'),
    dimensions: z.number().int().describe('Vector dimensions'),
  }).describe('Embedding information'),
});
export type StoreMemoryOutput = z.infer<typeof StoreMemoryOutputSchema>;

// ============================================================================
// Search Memories Tool Schemas
// ============================================================================

export const SearchMemoriesInputSchema = z.object({
  query: z.string().min(1).max(500).describe('Search query for semantic matching'),
  limit: z.number().int().min(1).max(50).default(10).describe('Maximum results to return'),
  minScore: z.number().min(0).max(1).default(0.5).describe('Minimum similarity score'),
  tags: z.array(z.string()).optional().describe('Filter by tags'),
  dateRange: z.object({
    start: z.string().datetime().optional().describe('Start date filter'),
    end: z.string().datetime().optional().describe('End date filter'),
  }).optional().describe('Date range filter'),
});
export type SearchMemoriesInput = z.infer<typeof SearchMemoriesInputSchema>;

export const MemoryResultSchema = z.object({
  id: z.string().describe('Memory ID'),
  content: z.string().describe('Memory content'),
  score: z.number().describe('Similarity score'),
  tags: z.array(z.string()).optional().describe('Memory tags'),
  createdAt: z.string().datetime().describe('Creation timestamp'),
});

export const SearchMemoriesOutputSchema = z.object({
  results: z.array(MemoryResultSchema).describe('Search results'),
  total: z.number().int().describe('Total matching memories'),
  searchTime: z.number().describe('Search time in milliseconds'),
});
export type SearchMemoriesOutput = z.infer<typeof SearchMemoriesOutputSchema>;

// ============================================================================
// Store Document Tool Schemas
// ============================================================================

export const StoreDocumentInputSchema = z.object({
  title: z.string().min(1).max(200).describe('Document title'),
  content: z.string().min(1).max(100000).describe('Document content'),
  type: z.enum(['text', 'markdown', 'code', 'structured']).default('text').describe('Document type'),
  metadata: z.record(z.any()).optional().describe('Additional metadata'),
  chunkSize: z.number().int().min(100).max(2000).default(500).describe('Chunk size for splitting'),
  chunkOverlap: z.number().int().min(0).max(500).default(50).describe('Overlap between chunks'),
});
export type StoreDocumentInput = z.infer<typeof StoreDocumentInputSchema>;

export const StoreDocumentOutputSchema = z.object({
  id: z.string().describe('Document ID'),
  title: z.string().describe('Document title'),
  chunks: z.number().int().describe('Number of chunks created'),
  totalTokens: z.number().int().describe('Estimated total tokens'),
  stored: z.boolean().describe('Whether storage was successful'),
});
export type StoreDocumentOutput = z.infer<typeof StoreDocumentOutputSchema>;

// ============================================================================
// Query Knowledge Tool Schemas
// ============================================================================

export const QueryKnowledgeInputSchema = z.object({
  subject: z.string().min(1).describe('Subject to query about'),
  relationTypes: z.array(z.string()).optional().describe('Filter by relation types'),
  depth: z.number().int().min(1).max(5).default(2).describe('Graph traversal depth'),
  includeProperties: z.boolean().default(true).describe('Include node properties'),
});
export type QueryKnowledgeInput = z.infer<typeof QueryKnowledgeInputSchema>;

export const KnowledgeNodeSchema = z.object({
  id: z.string().describe('Node ID'),
  label: z.string().describe('Node label/type'),
  properties: z.record(z.any()).optional().describe('Node properties'),
});

export const KnowledgeEdgeSchema = z.object({
  from: z.string().describe('Source node ID'),
  to: z.string().describe('Target node ID'),
  type: z.string().describe('Relationship type'),
  properties: z.record(z.any()).optional().describe('Edge properties'),
});

export const QueryKnowledgeOutputSchema = z.object({
  subject: z.string().describe('Queried subject'),
  nodes: z.array(KnowledgeNodeSchema).describe('Related nodes'),
  edges: z.array(KnowledgeEdgeSchema).describe('Relationships'),
  facts: z.array(z.string()).describe('Extracted facts'),
});
export type QueryKnowledgeOutput = z.infer<typeof QueryKnowledgeOutputSchema>;

// ============================================================================
// Error Types
// ============================================================================

export const GraphRAGErrorCodes = {
  MEMORY_NOT_FOUND: 'MEMORY_NOT_FOUND',
  DOCUMENT_TOO_LARGE: 'DOCUMENT_TOO_LARGE',
  EMBEDDING_FAILED: 'EMBEDDING_FAILED',
  SEARCH_FAILED: 'SEARCH_FAILED',
  GRAPH_QUERY_FAILED: 'GRAPH_QUERY_FAILED',
  SERVICE_UNAVAILABLE: 'SERVICE_UNAVAILABLE',
  QUOTA_EXCEEDED: 'QUOTA_EXCEEDED',
} as const;

export type GraphRAGErrorCode = keyof typeof GraphRAGErrorCodes;

File: templates/python/src/schemas.py
"""Pydantic schemas for plugin inputs and outputs.

These schemas define the structure and validation for all tool
inputs and outputs. They are automatically converted to JSON Schema
for the Plugin Intelligence Document.
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


# ============================================================================
# Enums
# ============================================================================


class ImportanceLevel(str, Enum):
    """Importance level for memories."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ErrorCode(str, Enum):
    """Error codes for plugin operations."""

    NOT_FOUND = "NOT_FOUND"
    VALIDATION_ERROR = "VALIDATION_ERROR"
    SERVICE_ERROR = "SERVICE_ERROR"
    QUOTA_EXCEEDED = "QUOTA_EXCEEDED"
    UNAUTHORIZED = "UNAUTHORIZED"


# ============================================================================
# Search Knowledge Schemas
# ============================================================================


class SearchKnowledgeInput(BaseModel):
    """Input schema for search_knowledge tool."""

    query: str = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="Search query for finding relevant knowledge",
    )
    limit: int = Field(
        default=10,
        ge=1,
        le=100,
        description="Maximum number of results to return",
    )
    min_score: float = Field(
        default=0.5,
        ge=0.0,
        le=1.0,
        description="Minimum similarity score threshold (0-1)",
    )
    tags: Optional[list[str]] = Field(
        default=None,
        description="Optional tags to filter results",
    )


class SearchResult(BaseModel):
    """A single search result."""

    id: str = Field(..., description="Unique identifier for the result")
    content: str = Field(..., description="The matched content")
    score: float = Field(..., ge=0, le=1, description="Similarity score")
    tags: list[str] = Field(default_factory=list, description="Associated tags")
    created_at: datetime = Field(..., description="When this was created")
    metadata: Optional[dict] = Field(default=None, description="Additional metadata")


class SearchKnowledgeOutput(BaseModel):
    """Output schema for search_knowledge tool."""

    results: list[SearchResult] = Field(..., description="List of search results")
    total: int = Field(..., ge=0, description="Total number of matches")
    query: str = Field(..., description="The original query")
    search_time_ms: int = Field(..., ge=0, description="Search duration in milliseconds")


# ============================================================================
# Store Memory Schemas
# ============================================================================


class StoreMemoryInput(BaseModel):
    """Input schema for store_memory tool."""

    content: str = Field(
        ...,
        min_length=1,
        max_length=10000,
        description="The memory content to store",
    )
    tags: list[str] = Field(
        default_factory=list,
        max_length=10,
        description="Tags for categorizing the memory",
    )
    importance: ImportanceLevel = Field(
        default=ImportanceLevel.MEDIUM,
        description="Importance level of the memory",
    )
    metadata: Optional[dict] = Field(
        default=None,
        description="Optional metadata to store with the memory",
    )


class EmbeddingInfo(BaseModel):
    """Information about the embedding."""

    model: str = Field(..., description="Embedding model used")
    dimensions: int = Field(..., description="Vector dimensions")


class StoreMemoryOutput(BaseModel):
    """Output schema for store_memory tool."""

    id: str = Field(..., description="Unique identifier for the stored memory")
    stored: bool = Field(..., description="Whether the memory was successfully stored")
    timestamp: datetime = Field(..., description="When the memory was stored")
    embedding: EmbeddingInfo = Field(..., description="Embedding information")


# ============================================================================
# Error Schemas
# ============================================================================


class PluginError(BaseModel):
    """Standard error response."""

    code: ErrorCode = Field(..., description="Error code")
    message: str = Field(..., description="Human-readable error message")
    details: Optional[dict] = Field(default=None, description="Additional error details")
    retryable: bool = Field(default=False, description="Whether the operation can be retried")
